Sources/Algorithms/LazySplitSequence.swift

```diff
@@ -0,0 +1,321 @@
+      maxSplits: Int,
+      omittingEmptySubsequences: Bool
+    ) {
+      self.base = base
+      self.isSeparator = whereSeparator
+      self.maxSplits = maxSplits
+      self.omittingEmptySubsequences = omittingEmptySubsequences
+    }
+  }
+}
+
+extension LazySplitSequence.Iterator: IteratorProtocol {
+  public mutating func next() -> Element? {
+    /// Separators mark the points where we want to split (cut in two) the base
+    /// collection, removing the separator in the process.
+    ///
+    /// Each split yields two subsequences, though splitting at the start or end
+    /// of a sequence yields an empty subsequence where there were no elements
+    /// adjacent to the cut.
+    ///
+    /// Thus the maximum number of subsequences returned after iterating the
+    /// entire base collection (including empty ones, if they are not omitted)
+    /// will be at most one more than the number of splits made (equivalently,
+    /// one more than the number of separators encountered).
+    ///
+    /// The number of splits is limited by `maxSplits`, and thus may be less
+    /// than the total number of separators in the base collection.
+    ///
+    ///     [1, 2, 42, 3, 4, 42, 5].split(separator: 42,
+    ///                                   omittingEmptySubsequences: false)
+    ///     // first split -> [1, 2], [3, 4, 42, 5]
+    ///     // last split  -> [1, 2], [3, 4], [5]
+    ///
+    ///     [1, 2, 42, 3, 4, 42, 5, 42].split(separator: 42,
+    ///                                       maxSplits: 2,
+    ///                                       omittingEmptySubsequences: false)
+    ///     // first split -> [1, 2], [3, 4, 42, 5, 42]
+    ///     // last split  -> [1, 2], [3, 4], [5, 42]
+    ///
+    ///     [42, 1, 42].split(separator: 42, omittingEmptySubsequences: false)
+    ///     // first split -> [], [1, 42]
+    ///     // last split  -> [], [1], []
+    ///
+    ///     [42, 42].split(separator: 42, omittingEmptySubsequences: false)
+    ///     // first split -> [], [42]
+    ///     // last split  -> [], [], []
+
+    var currentElement = base.next()
+    var subsequence: Element = []
+
+    while currentElement != nil {
+      if separatorCount < maxSplits && isSeparator(currentElement!) {
+        separatorCount += 1
+
+        if omittingEmptySubsequences && subsequence.isEmpty {
+          currentElement = base.next()
+          continue
+        } else {
+          break
+        }
+      } else {
+        subsequence.append(currentElement!)
+        currentElement = base.next()
+      }
+    }
+
+    if currentElement == nil {
+      if sequenceLength < separatorCount + 1 {
+        if !subsequence.isEmpty || !omittingEmptySubsequences {
+          sequenceLength += 1
+          return subsequence
+        } else {
+          return nil
+        }
+      } else {
+        return nil
+      }
+    }
+
+    defer {
+      sequenceLength += 1
+      subsequence = []
```

Comment:
If `subsequence` is now local there is not need for this to be here any more right?

Reply:
Absolutely. Thank you.